src/PlottingInterface.cc

```diff
@@ -0,0 +1,466 @@
+{
+  this->dataPtr->charts.insert(_chart);
+}
+
+//////////////////////////////////////////////////////
+void PlotData::RemoveChart(int _chart)
+{
+  auto chartIt = this->dataPtr->charts.find(_chart);
+  if (chartIt != this->dataPtr->charts.end())
+    this->dataPtr->charts.erase(chartIt);
+}
+
+//////////////////////////////////////////////////////
+int PlotData::ChartCount() const
+{
+  return this->dataPtr->charts.size();
+}
+
+//////////////////////////////////////////////////////
+std::set<int> &PlotData::Charts()
+{
+  return this->dataPtr->charts;
+}
+
+//////////////////////////////////////////////////////
+Topic::Topic(const std::string &_name) :
+    dataPtr(std::make_unique<TopicPrivate>())
+{
+    this->dataPtr->name = _name;
+}
+
+//////////////////////////////////////////////////////
+Topic::~Topic()
+{
+}
+
+//////////////////////////////////////////////////////
+std::string Topic::Name() const
+{
+  return this->dataPtr->name;
+}
+
+//////////////////////////////////////////////////////
+void Topic::Register(const std::string &_fieldPath, int _chart)
+{
+  // if a new field create a new field and register the chart
+  if (this->dataPtr->fields.count(_fieldPath) == 0)
+    this->dataPtr->fields[_fieldPath] = new PlotData();
+
+  this->dataPtr->fields[_fieldPath]->AddChart(_chart);
+}
+
+//////////////////////////////////////////////////////
+void Topic::UnRegister(const std::string &_fieldPath, int _chart)
+{
+  this->dataPtr->fields[_fieldPath]->RemoveChart(_chart);
+
+  // if no one registers to the field, remove it
+  if (!this->dataPtr->fields[_fieldPath]->ChartCount())
+    this->dataPtr->fields.erase(_fieldPath);
+}
+
+//////////////////////////////////////////////////////
+int Topic::FieldCount() const
+{
+  return this->dataPtr->fields.size();
+}
+
+//////////////////////////////////////////////////////
+std::map<std::string, PlotData*> &Topic::Fields()
+{
+  return this->dataPtr->fields;
+}
+
+//////////////////////////////////////////////////////
+void Topic::Callback(const google::protobuf::Message &_msg)
+{
+  for (auto fieldIt : this->dataPtr->fields)
+  {
+    auto msgDescriptor = _msg.GetDescriptor();
+    auto ref = _msg.GetReflection();
+
+    google::protobuf::Message *valueMsg = nullptr;
+
+    auto fieldFullPath = ignition::common::Split(fieldIt.first, '-');
+    int pathSize = fieldFullPath.size();
+
+    // loop until you reach the last field in the path
+    for (int i = 0; i < pathSize-1 ; i++)
+    {
+      std::string fieldName = fieldFullPath[i];
+
+      auto field = msgDescriptor->FindFieldByName(fieldName);
+
+      msgDescriptor = field->message_type();
+
+      if (valueMsg)
+        valueMsg = ref->MutableMessage
+                (const_cast<google::protobuf::Message *>(valueMsg), field);
+      else
+      {
+          valueMsg = ref->MutableMessage
+                  (const_cast<google::protobuf::Message *>(&_msg), field);
+      }
+
+        ref = valueMsg->GetReflection();
```

Comment:
oops, more bad indentation here and above

```suggestion
      ref = valueMsg->GetReflection();
```